### PR TITLE
Add instructions for how to use adversaries with Sabaki

### PR DESCRIPTION
**Pull request by ed1d1a8d**
I also created a public google drive folder to host our adversaries at https://drive.google.com/drive/folders/1-bGX-NQOh6MuRPoXJgYHb9-jWRJvviSg?usp=sharing.

### CODE REVIEW

File: configs/sabaki/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# How to set up sabaki
```
Comment (tomtseng) [inline configs/sabaki/README.md:1]: nit
```suggestion
# How to set up Sabaki
```

File: configs/sabaki/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+You can install Sabaki from https://sabaki.yichuanshen.de/.
+If you are on MacOS, you can install using homebrew with the command
```
Comment (tomtseng) [inline configs/sabaki/README.md:4]: nit
```suggestion
If you are on MacOS, you can install using Homebrew with the command
```

File: configs/sabaki/README.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+`brew install sabaki`.
+
+## Bot setup
+0. First build the C++ katago executable.
```
Comment (tomtseng) [inline configs/sabaki/README.md:8]: nit
```suggestion
0. First build the C++ KataGo executable.
```

File: configs/sabaki/README.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+   ```
+   Adjust the backend and whether to use tcmalloc as needed.
+1. Open up Sabaki preferences by pressing command comma (⌘,).
+2. Go to the Engines tab.
+3. A sample config would be the following:
+  ![bot-config-screenshot](bot-config-screenshot.png)
+  The first line is the bot name,
+  the second line is the path to the executable,
+  the third line are the flags passed to the executable,
+  are initial commands passed to gtp
```
Comment (tomtseng) [inline configs/sabaki/README.md:27]:
```suggestion
  and the fourth line is the initial commands passed to gtp
```

File: configs/sabaki/README.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+   make -j
+   ```
+   Adjust the backend and whether to use tcmalloc as needed.
+1. Open up Sabaki preferences by pressing command comma (⌘,).
+2. Go to the Engines tab.
+3. A sample config would be the following:
+  ![bot-config-screenshot](bot-config-screenshot.png)
+  The first line is the bot name,
+  the second line is the path to the executable,
+  the third line are the flags passed to the executable,
```
Comment (tomtseng) [inline configs/sabaki/README.md:26]:
```suggestion
  the third line is the flags passed to the executable,
```

File: configs/sabaki/README.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+   `engines/KataGo-custom/cpp`
+   and
+   `engines/KataGo-raw/cpp`
+   directories.
+   ```
+   cmake . -DUSE_BACKEND=OPENCL -DUSE_TCMALLOC=1
+   make -j
+   ```
+   Adjust the backend and whether to use tcmalloc as needed.
+1. Open up Sabaki preferences by pressing command comma (⌘,).
```
Comment (tomtseng) [inline configs/sabaki/README.md:20]: someone should check what the command is on Linux, but I'm also on macOS so I don't know

or we can just write that the user should go to Preferences in the top-level drop-down menu

File: configs/sabaki/gtp-v1.cfg [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+# Cache up to (2 ** this) many neural net evaluations in case of transpositions in the tree.
+# Uncomment and edit to change if you want to adjust a major component of KataGo's RAM usage.
+nnCacheSizePowerOfTwo = 20
+
+# Size of mutex pool for nnCache is (2 ** this).
+nnMutexPoolSizePowerOfTwo = 16
+
+numNNServerThreadsPerModel = 1
+openclDeviceToUseThread0 = 2
```
Comment (tomtseng) [inline configs/sabaki/gtp-v1.cfg:159]: is `= 2` always what we want here? I got an error with this set to `2`:
```
KataGo v1.10.0
Using koPOSITIONALscoreAREAtaxNONEsui1whbN rules initially, unless GTP/GUI overrides this
libc++abi: terminating with uncaught exception of type StringError: Requested gpuIdx/device 2 was not found, valid devices range from 0 to 1
```

File: configs/sabaki/gtp-adv600-victim1.cfg [hunk 8]
```diff
@@ -0,0 +1,9 @@
+@include gtp-v1.cfg
+
+numBots = 2
+maxVisits0 = 600
+searchAlgorithm0 = EMCTS1
+maxVisits1 = 1
+searchAlgorithm1 = MCTS
```
Comment (tomtseng) [inline configs/sabaki/gtp-adv600-victim1.cfg:7]: do we need `passingBehavior1 = avoid-pass-alive-territory`? (Depends on whether EMCTS invokes the victim's `Search::getPlaySelectionValues()` on victim nodes, I'm not familiar enough with the EMCTS code)

File: configs/sabaki/README.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+2. Go to the Engines tab.
+3. A sample config would be the following:
+  ![bot-config-screenshot](bot-config-screenshot.png)
+  The first line is the bot name,
+  the second line is the path to the executable,
+  the third line are the flags passed to the executable,
+  are initial commands passed to gtp
+  (in this case giving the bot infinite time to make moves).
+
```
Comment (tomtseng) [inline configs/sabaki/README.md:29]: now we've added an engine, but it may be helpful to add or link to instructions on how to actually use that engine or play against it

File: configs/sabaki/README.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+## List of bot configs
+
+Adversary models can be downloaded from
+https://drive.google.com/drive/folders/1-bGX-NQOh6MuRPoXJgYHb9-jWRJvviSg?usp=sharing.
+Base KataGo models can be downloaded from
+https://katagotraining.org/networks/.
+
```
Comment (tomtseng) [inline configs/sabaki/README.md:36]: would be nice to describe what each of these configs actually mean, especially how they relate to the adversaries/victims described in the paper. That could wait until a later, more final version of our paper though

**Review by tomtseng — COMMENTED** [review]
FWIW I followed the instructions to run Sabaki for the first time  and played a few moves against an adversary

File: configs/sabaki/README.md (hunk 6 above)
Comment (AdamGleave) [reply configs/sabaki/README.md:20]: In Linux it's File->Preferences

**Comment by AdamGleave** [comment]
Tom's already reviewed this so I'm not going to review it by default, but feel free to tag me if I should look at all/some of it.

**Comment by ed1d1a8d** [comment]
> @ed1d1a8d are you planning on making any changes to this, or is it just blocked on my/Tom's review?

Still planning on making some changes here. Will re-request review when ready.

File: configs/sabaki/README.md (hunk 6 above)
Comment (ed1d1a8d) [reply configs/sabaki/README.md]: Updated with instructions for Linux.

File: configs/sabaki/gtp-v1.cfg (hunk 7 above)
Comment (ed1d1a8d) [reply configs/sabaki/gtp-v1.cfg]: No longer relevant due to changing the gtp config to use mostly defaults.

File: configs/sabaki/README.md (hunk 9 above)
Comment (ed1d1a8d) [reply configs/sabaki/README.md]: Linked to youtube video.

File: configs/sabaki/gtp-adv600-victim1.cfg (hunk 8 above)
Comment (ed1d1a8d) [reply configs/sabaki/gtp-adv600-victim1.cfg]: We copy the config in https://github.com/AlignmentResearch/go_attack/blob/main/configs/gtp-amcts.cfg now, since this config was the one used for the 10 million visit benchmark and it worked pretty well.

File: sabaki/README.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+cyclic-adv-s545m-v600-vm-cp505-v1-s
+/Users/ttw/code/go_attack/engines/KataGo-custom/cpp/katago
+gtp -model /Users/ttw/code/go_attack/sabaki/models/adv/cyclic-adv-s545065216-d136760487.bin.gz -victim-model /Users/ttw/code/go_attack/sabaki/models/victims/kata1-b40c256-s11840935168-d2898845681.bin.gz -config /Users/ttw/code/go_attack/configs/gtp/gtp-adv600-vm1-s.cfg
+time_settings 0 1 0
+
+# Weaker cyclic-adversary (with dragonslayer strategy)
+cyclic-adv-s349m-v600-vm-cp505-v1-s
+/Users/ttw/code/go_attack/engines/KataGo-custom/cpp/katago
+gtp -model /Users/ttw/code/go_attack/sabaki/models/adv/cyclic-adv-s349284096-d87808728.bin.gz -victim-model /Users/ttw/code/go_attack/sabaki/models/victims/kata1-b40c256-s11840935168-d2898845681.bin.gz -config /Users/ttw/code/go_attack/configs/gtp/gtp-adv600-vm1-s.cfg
+time_settings 0 1 0
```
Comment (tomtseng) [inline sabaki/README.md:54]: Are we actually still interested in this weaker adversary or should we just delete it from here?

File: sabaki/README.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+# Pass-trick adversary
```
Comment (tomtseng) [inline sabaki/README.md:56]: nit: for consistency with paper, rename `Pass-trick adversary` to `pass-adversary`?

File: sabaki/README.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+time_settings 0 1 0
+
+# Cyclic-adversary over ssh
+ssh-adv505h-s545m-v600-vm-cp505-v1-s
+ssh
+rnn -tt 'bash -l -c "/nas/ucb/tony/go-attack/gtp-host/go_attack/sabaki/scripts/docker-katago-custom.sh gtp -model /models/adv/cyclic-adv-s545065216-d136760487.bin.gz -victim-model /models/victims/kata1-b40c256-s11840935168-d2898845681.bin.gz -config /go_attack/configs/gtp/gtp-adv600-vm1-s.cfg"'
+time_settings 0 1 0
+```
+
+Also see https://github.com/AlignmentResearch/KataGoVisualizer/blob/3216f9b5bebaf1377dc1e320294abcc0f65bd8bd/notebooks/notebooks/paper/estimate-flops-adv.ipynb for locations of adversaries on the nas.
```
Comment (tomtseng) [inline sabaki/README.md:81]: this line will be confusing to people besides us, should we just remove it?

File: sabaki/scripts/docker-katago-raw.sh [hunk 14]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+cd ${SCRIPT_DIR}
+
+GIT_ROOT=$(git rev-parse --show-toplevel)
+cd ${GIT_ROOT}
+
+docker run \
+    -v sabaki/models:/models \
```
Comment (tomtseng) [inline sabaki/scripts/docker-katago-raw.sh:10]: does this not need a `--gpus device=0`?

looks like `docker-katago-raw.sh` and `docker-katago-custom.sh` could be combined into one file for cleanliness if you want to be diligent — the only difference is calling  `/engines/KataGo-custom/cpp/katago` vs. `/engines/KataGo-raw/cpp/katago`

File: sabaki/README.md [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+# Latest with no search
+cp505-v1
+/Users/ttw/code/go_attack/engines/KataGo-raw/cpp/katago
+gtp -model /Users/ttw/code/go_attack/sabaki/models/victims/kata1-b40c256-s11840935168-d2898845681.bin.gz -config /Users/ttw/code/go_attack/configs/gtp/gtp-v1.cfg
+time_settings 0 1 0
+
+# Cyclic-adversary over ssh
+ssh-adv505h-s545m-v600-vm-cp505-v1-s
+ssh
+rnn -tt 'bash -l -c "/nas/ucb/tony/go-attack/gtp-host/go_attack/sabaki/scripts/docker-katago-custom.sh gtp -model /models/adv/cyclic-adv-s545065216-d136760487.bin.gz -victim-model /models/victims/kata1-b40c256-s11840935168-d2898845681.bin.gz -config /go_attack/configs/gtp/gtp-adv600-vm1-s.cfg"'
```
Comment (tomtseng) [inline sabaki/README.md:77]: what happens if we don't use the `-l` flag on `bash`?

File: sabaki/scripts/ssh-katago-custom.sh [hunk 16]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+bash_command="/nas/ucb/tony/go-attack/gtp-host/go_attack/sabaki/scripts/docker-katago-custom.sh $@"
+echo ${bash_command}
+ssh rnn -tt 'bash -l -c "'${bash_command}'"'
```
Comment (tomtseng) [inline sabaki/scripts/ssh-katago-custom.sh:4]: What is this script used for?

File: configs/gtp/gtp-adv600-vm1-s.cfg [hunk 17]
```diff
@@ -0,0 +1,9 @@
+@include gtp-v1.cfg
```
Comment (tomtseng) [inline configs/gtp/gtp-adv600-vm1-s.cfg:1]: should we move our existing configs `gtp-amcts.cfg`, `gtp-base.cfg`, `gtp-raw.cfg` into `configs/gtp`? Or just rename `configs/gtp` to `configs/sabaki`?
I'm also wondering whether `gtp-v1.cfg` should import `gtp-base.cfg` and make modifications from there

File: sabaki/scripts/ssh-katago-custom.sh (hunk 16 above)
Comment (ed1d1a8d) [reply sabaki/scripts/ssh-katago-custom.sh:4]: Ah just an old file that was previously used but not anymore since I put all the commands into Sabaki directly. Will delete.

File: sabaki/README.md (hunk 15 above)
Comment (ed1d1a8d) [reply sabaki/README.md]: Without the `-l` flag, you get the error:
```
Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
docker: invalid reference format: repository name must be lowercase.
See 'docker run --help'.
```
This is because /nas/ucb/tony/go-attack/gtp-host/go_attack/sabaki/scripts/docker-katago-custom.sh tries to build and launch a docker container, which for some reason doesn't work without `-l`.

File: sabaki/README.md (hunk 13 above)
Comment (ed1d1a8d) [reply sabaki/README.md]: Moved to separate section marked as internal

File: configs/gtp/gtp-adv600-vm1-s.cfg (hunk 17 above)
Comment (ed1d1a8d) [reply configs/gtp/gtp-adv600-vm1-s.cfg:1]: Moved and refactored configs.

File: sabaki/scripts/docker-katago-raw.sh (hunk 14 above)
Comment (ed1d1a8d) [reply sabaki/scripts/docker-katago-raw.sh]: Good suggestion, have combined the two scripts.

**Comment by ed1d1a8d** [comment]
Thanks for the review Tom. I made your requested changes and also updated the main README to be a bit more clean and to link to the sabaki instructions. Re-requesting review.

File: README.md [hunk 18]
```diff
@@ -2,20 +2,34 @@
...
+View our website here: https://goattack.far.ai/.
+
+To run our adversary with Sabaki, see [this guide](sabaki/README.md).
+
+# Development / testing information
+
+To clone this repository,
+run one of the following commands
+```
+# Via https
```
Comment (tomtseng) [inline README.md:16]: silly nit:
```suggestion
# Via HTTPS
```

File: README.md [hunk 19]
```diff
@@ -44,30 +58,9 @@ parameters of the run (
...
-        - `cd go_attack`
-    - Build C++ and Python Docker containers, and run the C++ container. A KataGo executable can be found in the `/engines/KataGo-custom/cpp` directory inside the container.
-        - `docker build . -f compose/cpp/Dockerfile -t humancompatibleai/goattack:cpp`
-        - `docker build . -f compose/python/Dockerfile -t humancompatibleai/goattack:python`
-        - `docker run --gpus all -v ~/go_attack:/go_attack -it humancompatibleai/goattack:cpp`
-    - Download model weights
-        - `cd /go_attack/configs/katago && wget -i model_list.txt -P /go_attack/victim-models`
-    - Test if the installation is successful
-        - `cd /engines/KataGo-custom/cpp/ && ./katago benchmark -model /go_attack/victim-models/g170-b40c256x2-s5095420928-d1229425124.bin.gz -config /go_attack/configs/katago/baseline_attack.cfg`
+## Website and analysis notebooks.
```
Comment (tomtseng) [inline README.md:61]:
```suggestion
## Website and analysis notebooks
```

File: README.md [hunk 20]
```diff
@@ -2,20 +2,34 @@
...
-Modifications to KataGo *are not* tracked in this repository and should instead be made to the [HumanCompatibleAI/KataGo-custom](https://github.com/HumanCompatibleAI/KataGo-custom) repository. We use code from KataGo-custom in this repository via Git submodules.
+Modifications to KataGo *are not* tracked in this repository and should instead be made to the [AlignmentResearch/KataGo-custom](https://github.com/AlignmentResearch/KataGo-custom) repository. We use code from KataGo-custom in this repository via Git submodules.
 
 KataGo-custom has the following significant branches:
 
 - `KataGo-custom/stable` contains our changes to the stable version of KataGo.
 - `KataGo-custom/master` tracks https://github.com/lightvector/KataGo.
 
-# Development / testing
+## Other setup
```
Comment (tomtseng) [inline README.md:32]: maybe: remove this section header and move the contents of this section above `##Git submodule: KataGo-custom`

File: README.md [hunk 21]
```diff
@@ -44,30 +58,9 @@ parameters of the run (
...
-    - Agent code: https://github.com/HumanCompatibleAI/KataGo-custom
-    - Controller code: https://github.com/HumanCompatibleAI/gogui
-- **Setting up**
-    - Download the repo
-        - `git clone --recurse-submodules https://github.com/HumanCompatibleAI/go_attack.git`
-        - `cd go_attack`
-    - Build C++ and Python Docker containers, and run the C++ container. A KataGo executable can be found in the `/engines/KataGo-custom/cpp` directory inside the container.
-        - `docker build . -f compose/cpp/Dockerfile -t humancompatibleai/goattack:cpp`
-        - `docker build . -f compose/python/Dockerfile -t humancompatibleai/goattack:python`
-        - `docker run --gpus all -v ~/go_attack:/go_attack -it humancompatibleai/goattack:cpp`
```
Comment (tomtseng) [inline README.md:66]: I think this `docker run` command and the statement "A KataGo executable can be found in the `/engines/KataGo-custom/cpp` directory inside the [C++] container" are still useful.

Either move the useful information here to the `## Individual containers` section, or add it inside the header comment of the C++ Dockerfile?

File: configs/gtp-base.cfg [hunk 22]
```diff
@@ -1,5 +1,3 @@
-@include compute/1gpu.cfg
```
Comment (tomtseng) [inline configs/gtp-base.cfg:1]: what's the reason for this change? was including `compute/1gpu.cfg` messing up Sabaki?

File: sabaki/README.md [hunk 23]
```diff
@@ -36,46 +36,55 @@ The base KataGo models are downloaded from https://katagotraining.org/networks/.
...
+gtp -model /Users/ttw/code/go_attack/sabaki/models/victims/kata1-b40c256-s11840935168-d2898845681.bin.gz -config /Users/ttw/code/go_attack/configs/sabaki/gtp-v1.cfg
 time_settings 0 1 0
+```
 
+## Internal notes
+The following config enables playing against an adversary (or victim) over ssh.
+To do this on a machine M without /nas/ access,
+you need to clone the go-attack repo on M,
+download the adversary and victim models on M,
+and make sure docker works on M.
```
Comment (tomtseng) [inline sabaki/README.md:81]: nits:
```suggestion
The following config enables playing against an adversary (or victim) over SSH.
To do this on a machine M without `/nas/` access,
you need to clone the `go_attack` repo on M,
download the adversary and victim models on M,
and make sure Docker works on M.
```

File: sabaki/scripts/docker-katago.sh [hunk 24]
```diff
@@ -6,8 +6,13 @@ cd ${SCRIPT_DIR}
 GIT_ROOT=$(git rev-parse --show-toplevel)
 cd ${GIT_ROOT}
 
+# Get which submodule to use (raw or custom)
+KATAGO_TYPE=$1; shift
+
+# Build and run the docker image
+# Manually change the assigned gpu devices as needed.
```
Comment (tomtseng) [inline sabaki/scripts/docker-katago.sh:13]: nits
```suggestion
# Build and run the Docker image
# Manually change the assigned GPU devices as needed.
```

File: configs/sabaki/gtp-v1.cfg [hunk 25]
```diff
@@ -0,0 +1,9 @@
+@include ../gtp-base.cfg
+
+maxVisits = 1
+
+# Set logSearchInfo to true for debugging!
+# Setting to true will break playing over ssh however!
```
Comment (tomtseng) [inline configs/sabaki/gtp-v1.cfg:6]: silly nits:
```suggestion
# Set logSearchInfo to true for debugging.
# Setting to true will break playing over SSH, however!
```

**Review by tomtseng — APPROVED** [review]

File: configs/gtp-base.cfg (hunk 22 above)
Comment (ed1d1a8d) [reply configs/gtp-base.cfg:1]: Yeah the sabaki configs are designed to work with cpus only.

File: README.md (hunk 21 above)
Comment (ed1d1a8d) [reply README.md:66]: Added back in.

File: README.md (hunk 20 above)
Comment (ed1d1a8d) [reply README.md]: Moved higher.
